### PR TITLE
FOTA Service

### LINKED ISSUE #46: Service Proposal: FOTAService
### BLE Standard Service:

- [ ] Yes
- [X] No

### Proposed UUID(s) or link to specification document:

See specification and documentation on fota-service branch here:

https://github.com/ARMmbed/mbed-os-experimental-ble-services/tree/5328cacea69452901a691b7380366b6a188cf503/services/FOTA

@pan- @noonfom @paul-szczepanek-arm I took aspects from our discussion in #13 and wrote up a more formal specification README, taking the LinkLossService as an example.

The newly-renamed `FOTAService` strips out some complexity from my original DFUService proposal. I think the new specification covers the most basic use cases while providing for more advanced features and extensions in the future. 

I have yet to refactor the code, but I'm planning to split off the `BlockDevice` dependency into standard `FOTAService::EventHandlers` that we provide. A user can implement their own custom logic if they wish, but we can provide a basic event handler for writing to a given block device. This also serves as a way for us to provide modular extensions (eg: delta updates, trigger connection parameter updates, etc) in the future as part of our standard library of event handlers.

Before I go much further on the code and tests, I wanted to get your feedback.

Let me know what you think and if I missed any corner cases.

### PR DESCRIPTION
**Pull request by AGlass0fMilk**
This PR contains the FOTA service and associated tests as outlined in #46.

Closes #46

### CODE REVIEW

**Comment by AGlass0fMilk** [comment]
After working with this service on some projects, I think we should introduce another standard FOTA op-code, "FOTA confirm".

Bootloaders like Mcuboot support rolling back firmware updates if they do not mark themselves as "good". ie: the firmware has a severe bug that prevents normal operation. When reset, mcuboot will then revert the update to prevent "bricking".

One possible failure of this system is, since the update mechanism is in the application, there is a possibility a fatal bug could be introduced into the update mechanism code. This would prevent any future firmware upgrades to the device. The FOTA Confirm op-code would allow the FOTA client to mark the updated firmware as "OK".

This is essentially a self-test of the firmware update mechanism in the BLE FOTA service case -- if BLE can be brought up, a connection can be made, and a characteristic can be written, it is pretty safe to assume that the FOTA code has not been broken by the update.

There are some other recommended self-tests that should be performed, such as testing update flash memory initialization.

I think this would be a common enough use case to deserve a dedicated, standard op-code.

@pan- Thoughts on this? This is a problem for all field-upgradeable devices where the update mechanism is in the application itself. If Mbed-OS creates a framework for DFU, it should include features like a flexible, easy-to-integrate "factory reset" that can reinstall an image stored on a dedicated "recovery partition" (eg: a well-tested application that does nothing but let the firmware be reprogrammed in the case of a broken update).

File: .github/workflows/run_tests.yml [hunk 1]
```diff
@@ -56,3 +56,4 @@ jobs:
           ./tests/TESTS/build.sh -s DeviceInformation -t GCC_ARM -m NRF52840_DK
           ./tests/TESTS/build.sh -s LinkLoss -t GCC_ARM -m NRF52840_DK
           ./tests/TESTS/build.sh -s LinkLoss -t GCC_ARM -m DISCO_L496AG
+          ./tests/TESTS/build.sh -s FOTA -t GCC_ARM -m NRF52840_DK
```
Comment (noonfom) [inline .github/workflows/run_tests.yml:59]: It should also work on disco. There are MCUboot problems but the [mock](https://github.com/noonfom/mbed-os-example-ble/tree/mcuboot/BLE_GattServer_FOTAService/Mock) example works fine locally, i.e. I can transfer a binary into the external flash of my [B-L4S5I-IOT01A](https://os.mbed.com/platforms/B-L4S5I-IOT01A/) using the FOTA service.

**Note:** The disco build is also missing for the Device Information service: #62

File: common/inc/BaseService.h [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+//
+//    BaseService();
+//
+//
+//
+//
+//};
+
+
+#endif /* MBED_OS_EXPERIMENTAL_BLE_SERVICES_INC_BASESERVICE_H_ */
```
Comment (noonfom) [inline common/inc/BaseService.h:40]: It's a good idea but perhaps we should add it in a separate PR with the other utility files:

* ConnectionThingy.h
* GattPresentationFormatDescriptor.h
* CharacteristicPresentationFormatDescriptor.h
* CharacteristicUserDescriptionDescriptor.h

File: common/inc/ConnectionThingy.h [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+/**
+ * Connection... Thingy?
+ * A common problem encountered when developing BLE applications is:
+ *
+ * Characteristic read/write/notify/indicate events are handled in batch.
+ *
+ * This creates a lot of boilerplate code where a service must filter out by comparing
+ * which handle was written to the callback's parameters, figure out which connection,
+ * and then perform the appropriate action.
```
Comment (noonfom) [inline common/inc/ConnectionThingy.h:31]: Do [`setWriteAuthorizationCallback`](https://github.com/ARMmbed/mbed-os/blob/aeaac0e70c76fbc5b753ee8c836e369edf1a7fb4/connectivity/FEATURE_BLE/include/ble/gatt/GattCharacteristic.h#L1556) and [`setReadAuthorizationCallback`](https://github.com/ARMmbed/mbed-os/blob/aeaac0e70c76fbc5b753ee8c836e369edf1a7fb4/connectivity/FEATURE_BLE/include/ble/gatt/GattCharacteristic.h#L1603) not solve this problem?

File: common/inc/ConnectionThingy.h [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+ *
+ * Maybe connection oriented service? Service logic
+ *
+ *
+ */
+
+
+
+
+#endif /* MBED_OS_EXPERIMENTAL_BLE_SERVICES_COMMON_INC_CONNECTIONTHINGY_H_ */
```
Comment (noonfom) [inline common/inc/ConnectionThingy.h:61]: See comment under BaseService.h

File: common/inc/GattPresentationFormatDescriptor.h [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+	private:
+
+		// Wouldn't it be nice if packing structs was more consistently supported by compilers?
+		//struct GattCharacteristic::PresentationFormat_t format;
+
+		// In lieu of using the struct above, packing issues makes us have to use a raw buffer
+		uint8_t format[7];
+};
+
+#endif /* EP_CORE_FEATURES_FEATURE_BLE_GATTPRESENTATIONFORMATDESCRIPTOR_H_ */
```
Comment (noonfom) [inline common/inc/GattPresentationFormatDescriptor.h:70]: See comment under BaseService.h

File: descriptors/CharacteristicPresentationFormatDescriptor.h [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    private:
+
+        // Wouldn't it be nice if packing structs was more consistently supported by compilers?
+        //struct GattCharacteristic::PresentationFormat_t format;
+
+        // In lieu of using the struct above, packing issues makes us have to use a raw buffer
+        uint8_t format[7];
+};
+
+#endif /* MBED_OS_EXPERIMENTAL_BLE_SERVICES_DESCRIPTORS_CHARACTERISTICPRESENTATIONFORMATDESCRIPTOR_H_ */
```
Comment (noonfom) [inline descriptors/CharacteristicPresentationFormatDescriptor.h:94]: See comment under BaseService.h

File: descriptors/CharacteristicUserDescriptionDescriptor.h [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+private:
+
+    const char* _user_description;
+
+
+};
+
+#endif /* MBED_OS_EXPERIMENTAL_BLE_SERVICES_DESCRIPTORS_CHARACTERISTICUSERDESCRIPTIONDESCRIPTOR_H_ */
```
Comment (noonfom) [inline descriptors/CharacteristicUserDescriptionDescriptor.h:59]: See comment under BaseService.h

File: scripts/bootstrap.sh [hunk 8]
```diff
@@ -13,59 +13,56 @@
...
 
 # Add symlinks
-symlink dependencies/mbed-os       tests/UNITTESTS/mbed-os
-
-symlink dependencies/mbed-os       tests/TESTS/LinkLoss/device/mbed-os
-symlink services/LinkLoss          tests/TESTS/LinkLoss/device/LinkLoss
+symlink "$ROOT"/dependencies/mbed-os "$ROOT"/tests/UNITTESTS/mbed-os
 
-symlink dependencies/mbed-os       tests/TESTS/DeviceInformation/device/mbed-os
-symlink services/DeviceInformation tests/TESTS/DeviceInformation/device/DeviceInformation
```
Comment (noonfom) [inline scripts/bootstrap.sh:39]: Both files are required to build the integration tests for the Device Information service. Can you add them back in?

File: scripts/bootstrap.sh [hunk 9]
```diff
@@ -13,59 +13,56 @@
...
+symlink "$ROOT"/dependencies/mbed-os "$ROOT"/tests/TESTS/FOTA/device/mbed-os
+symlink "$ROOT"/common               "$ROOT"/tests/TESTS/FOTA/device/common
+symlink "$ROOT"/descriptors          "$ROOT"/tests/TESTS/FOTA/device/descriptors
+symlink "$ROOT"/services/FOTA        "$ROOT"/tests/TESTS/FOTA/device/FOTA
 
-# Create mbed-os.lib for CMake builds
-echo "https://github.com/ARMmbed/mbed-os" > tests/TESTS/LinkLoss/device/mbed-os.lib
-echo "https://github.com/ARMmbed/mbed-os" > tests/TESTS/DeviceInformation/device/mbed-os.lib
```
Comment (noonfom) [inline scripts/bootstrap.sh:43]: Mbed CLI 2 requires that a `mbed-os.lib` be present in the project root.

The following is also required for the FOTA service:

```shell
echo "https://github.com/ARMmbed/mbed-os" > tests/TESTS/FOTA/device/mbed-os.lib
```

File: tests/TESTS/FOTA/device/.mbed [hunk 10]
```diff
@@ -0,0 +1 @@
+ROOT=.
```
Comment (noonfom) [inline tests/TESTS/FOTA/device/.mbed:1]: hmm... I don't think we should have .mbed files in device folders.

File: tests/TESTS/FOTA/device/mbed-os.lib [hunk 11]
```diff
@@ -0,0 +1 @@
+https://github.com/ARMmbed/mbed-os
```
Comment (noonfom) [inline tests/TESTS/FOTA/device/mbed-os.lib:1]: That should be added during bootstrap.

File: common/inc/BaseService.h (hunk 2 above)
Comment (AGlass0fMilk) [reply common/inc/BaseService.h:40]: Ha, sorry, these files weren't supposed to be included in this PR :sweat_smile:

File: common/inc/ConnectionThingy.h (hunk 3 above)
Comment (AGlass0fMilk) [reply common/inc/ConnectionThingy.h:31]: Unrelated the PR at hand. Those APIs "solve" the problem, but I think it creates a confusing situation in the BLE API:
You can still handle characteristic writes that are authorized but you then have to filter them based on the value handle. It creates two ways to do the same thing which can be confusing to users.

The naming of "setWriteAuthorizationCallback" and "setReadAuthorizationCallback" imply they are _only_ for read authorization/write authorization. Likewise, the naming of `GattServer::EventHandler::onCharacteristicWritten` (or whatever it is) implies it should be where characteristic writes are handled.
